### DISCUSSION TITLE
Options1 question

### DISCUSSION THREAD

**Discussion by Fr3027** [rust-lang/rustlings#2035] [Q&A] [answered]
```
fn maybe_icecream(time_of_day: u16) -> Option<u16> {
    if time_of_day<22{
        Some(5)
    }
    if time_of_day<24{
        Some(0)
    }
    None
}
```
The compiler throws an error: "expected `()`, found enum `Option`".
My point is if any of the conditions match then the function will return Some(), if not, the function will return None by default. 
This style is supported by java, which I can't understand why rust not, help me please!

**Comment by aleksanderkrauze** [comment, marked as answer]
You can either do

```rust
fn maybe_icecream(time_of_day: u16) -> Option<u16> {
    if time_of_day<22{
        return Some(5);
    }
    if time_of_day<24{
        return Some(0);
    }
    None
}
```

or

```rust
fn maybe_icecream(time_of_day: u16) -> Option<u16> {
    if time_of_day<22{
        Some(5)
    } else if time_of_day<24{
        Some(0)
    } else {
        None
    }
}
```

The compiler even suggests the first solution:

```
help: you might have meant to return this value
  |
3 |         return Some(5);
  |         ++++++        +
```

The reason for this is that in rust `if` statements are [expressions](https://doc.rust-lang.org/reference/expressions/if-expr.html). In your example those `if` branches are **unrelated** to each other. Therefore rust looks at them in isolation and treats them like **statements** which must return unit type `()`. You can solve this in two ways.

You can add `return` keywords (like compiler suggests, also my first example). This will make if expression return `()` as they should[^1].

Or you can combine three branches in one expression (like in my second example). Then each branch of this expression will return `Option<u16>` so the whole expression will evaluate to `Option<u16>` and this will be returned from your function.

[^1]: In reality this is more subtle. `return` is also an expression that returns [never](https://doc.rust-lang.org/std/primitive.never.html) type, which will then be [coerced](https://doc.rust-lang.org/reference/type-coercions.html) into unit type.
